src/main/java/io/jenkins/plugins/extendedsecuritysettings/UnauthorizedUserHttpHeaderRestrictionFilter.java

```diff
@@ -0,0 +1,155 @@
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Restricted(NoExternalUse.class)
+public class UnauthorizedUserHttpHeaderRestrictionFilter implements Filter {
+
+    @Initializer
+    public static void initialize() throws ServletException {
+        PluginServletFilter.addFilter(new UnauthorizedUserHttpHeaderRestrictionFilter());
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        HttpServletResponseWrapper responseWrapper = null;
+        if (isFilterableResponse(response) && !hasOverallRead()) {
+            Set<String> headerNames = getFilteredHeaderNames();
+            if (!headerNames.isEmpty()) {
+                if (response instanceof HttpServletResponseWrapper) {
+                    // reuse existing wrapper to ensure proper handling of AccessDeniedException
+                    responseWrapper = (HttpServletResponseWrapper) response;
+                    ServletResponse wrappedResponse = responseWrapper.getResponse();
+                    if (isFilterableResponse(wrappedResponse)) {
+                        responseWrapper.setResponse(new HeaderRemovingResponseWrapper((HttpServletResponse) wrappedResponse, headerNames));
+                    }
+                } else {
+                    responseWrapper = new HeaderRemovingResponseWrapper((HttpServletResponse) response, headerNames);
+                }
+            }
+        }
+        chain.doFilter(request, responseWrapper != null ? responseWrapper : response);
+    }
+
+    private static boolean isFilterableResponse(@Nonnull ServletResponse response) {
+        return response instanceof HttpServletResponse && !response.isCommitted();
+    }
+
+    private static boolean hasOverallRead() {
+        Jenkins jenkins = Jenkins.getInstanceOrNull();
+        return jenkins != null && jenkins.hasPermission(Jenkins.READ);
+    }
+
+    private static @Nonnull Set<String> getFilteredHeaderNames() {
+        Set<HttpHeaderName> headerNames = ExtendedSecuritySettings.get().getHttpHeaderNames();
+        if (headerNames == null || headerNames.isEmpty()) {
+            return Collections.emptySet();
+        }
+        return headerNames.stream()
+                .map(HttpHeaderName::getHeaderName)
+                .filter(headerName -> headerName != null && !headerName.isEmpty())
```

Comment:
As you like Lambdas you could use `StringUtils::isNotEmpty` to achieve the same (+ the import)